File: java/src/mixto/Mixto.java
package mixto;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import okhttp3.*;

import java.io.IOException;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

// TODO 🔥 update for v2

public final class Mixto {
    private final String MIXTO_API_KEY, MIXTO_HOST, MIXTO_WORKSPACE;
    private final HashMap<String, String> EMPTY_QUERY = new HashMap<>();
    private final OkHttpClient client = new OkHttpClient();

    /**
     * @param apiKey    Mixto API key
     * @param host      Mixto host URL
     * @param workspace Current workspace
     * @throws IllegalArgumentException Throws error if API key or Host is not defined
     */
    public Mixto(String apiKey, String host, String workspace) throws IllegalArgumentException {
        this.MIXTO_API_KEY = apiKey;
        this.MIXTO_HOST = host;
        this.MIXTO_WORKSPACE = workspace;
        if (apiKey.equals("") || host.equals("")) {
            throw new IllegalArgumentException("Missing required parameters");
        }
    }

    /**
     * Create a new Mixto instance from the local config file. If valid environment variables are set,
     * this class is initialized with the environment variables instead.
     *
     * @throws Exception raised if local config file cannot be read, or parsing fails
     */
    public Mixto() throws Exception {
        /* Check envars for values first */
        var envApiKey = System.getenv("MIXTO_API_KEY");
        var envHost = System.getenv("MIXTO_HOST");
        var envWorkspace = System.getenv("MIXTO_WORKSPACE");
        if (envApiKey == null || envHost == null || envWorkspace == null) {
            var configFile = System.getProperty("user.home") + "/.mixto.json";
            var configMapper = new ObjectMapper();
            configMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
            var config = configMapper.readValue(Paths.get(configFile).toFile(), Config.class);
            this.MIXTO_API_KEY = config.api_key;
            this.MIXTO_HOST = config.host;
            this.MIXTO_WORKSPACE = config.workspace_id;
        } else {
            this.MIXTO_API_KEY = envApiKey;
            this.MIXTO_HOST = envHost;
            this.MIXTO_WORKSPACE = envWorkspace;
        }
    }

    /**
     * Make a new URL with the provided path and host from config
     * @param path to join
     * @return HttpUrl.Builder object
     */
    private HttpUrl.Builder makeURL(String path) {
        return HttpUrl.parse(this.MIXTO_HOST).newBuilder().addPathSegments(path);
    }

    /**
     * Make a request object
     * @param endpoint for api
     * @param query params. Can be empty
     * @return Request.Builder object
     */
    private Request.Builder makeRequestObject(String endpoint, HashMap<String, String> query) {
        /* Replace forward slash from endpoint */
        endpoint = endpoint.replaceFirst("^/", "");
        /* Build the request URL */
        var url = this.makeURL(endpoint);
        /* Add query parameters */
        for (Map.Entry<String, String> q : query.entrySet()) {
            url.addQueryParameter(q.getKey(), q.getValue());
        }
        var apiUrl = url.build().url();
        /* Make request object */
        var request = new Request.Builder().url(apiUrl);
        request.header("x-api-key", this.MIXTO_API_KEY);
        return request;

    }

    /**
     * Check is status code is 200
     * @param response object
     * @return true if not 200
     */
    private boolean hasBadStatus(Response response) {
        return response.code() != 200;
    }

    /**
     * Make a GET request
     * @param endpoint to the api endpoint
     * @param query any query parameters
     * @return Response object
     * @throws Exception If status code is not 200
     */
    public Response MakeRequest(String endpoint, HashMap<String, String> query) throws Exception {
        var request = this.makeRequestObject(endpoint, query);
        var response = client.newCall(request.build()).execute();
        if (response.code() != 200)
            throw new IOException("Bad status code: " + response.code());
        return response;
    }

    /**
     * Make requests that are not GET requests. Use `jsonfromString` to easily build the optional json body
     *
     * @param method for request
     * @param endpoint for api request
     * @param body request json body
     * @return Response object
     * @throws Exception If status is not 200
     */
    public Response MakeRequest(String method, String endpoint, RequestBody body) throws Exception {
        var request = this.makeRequestObject(endpoint, EMPTY_QUERY);
        var response = client.newCall(request.method(method, body).build()).execute();
        if (hasBadStatus(response))
            throw new IOException("Bad status code: " + response.code());
        return response;
    }

    /**
     * Get an array of all workspaces
     *
     * @return Workspace[]
     * @throws Exception exception
     */
    public Workspace[] GetWorkspaces() throws Exception {
        var query = new HashMap<String, String>();
        var responseMapper = new ObjectMapper();
        var response = this.MakeRequest("/api/v1/workspace", query);
        Workspace[] workspaces = responseMapper.readValue(response.body().byteStream(), Workspace[].class);
        if (MIXTO_WORKSPACE == null) {
            return new Workspace[0];
        }
        return workspaces;
    }

    public Entry[] GetEntries(Boolean includeCommits) throws Exception {
        var body = new HashMap<String, Object>();
        var responseMapper = new ObjectMapper();
        body.put("workspace_id", this.MIXTO_WORKSPACE);
        body.put("include_commits", includeCommits);
        var jsonBody = new ObjectMapper().writeValueAsString(body);
        var resp = this.MakeRequest("POST", "/api/v1/workspace", jsonfromString(jsonBody));
        assert resp.body() != null;
        return responseMapper.readValue(resp.body().byteStream(), Entry[].class);
    }

    /**
     * Add a commit to Mixto
     *
     * @param data    to commit
     * @param entryID id
     * @param title   for commit
     * @return Response object
     * @throws Exception if status code is not 200
     */
    public Response AddCommit(String data, String entryID, String title) throws Exception {
        var entryURL = "/api/v1/commit";
        var body = new HashMap<String, String>();
        body.put("commit_type", "tool");
        body.put("title", title);
        body.put("data", data);
        body.put("entry_id", entryID);
        body.put("workspace_id", this.MIXTO_WORKSPACE);
        var jsonBody = new ObjectMapper().writeValueAsString(body);
        return MakeRequest("POST", entryURL, jsonfromString(jsonBody));
    }

    /**
     * Convert JSON string to RequestBody for non GET requests
     * @param jsonString json as string
     * @return RequestBody to be used in MakeRequest method
     */
    public RequestBody jsonfromString(String jsonString) {
        var jsonContentTypeHeader = MediaType.parse("application/json; charset=utf-8");
        return RequestBody.create(jsonContentTypeHeader, jsonString);
    }
}

File: python3/mixto.py
# Mixto lite lib for python3

from typing import Dict, List, Union, Any
from urllib.request import Request, urlopen
from urllib.parse import urlencode, urljoin
from urllib.error import HTTPError
from pathlib import Path
from os import getenv
import json

MIXTO_ENTRY_ID = getenv("MIXTO_ENTRY_ID")
MIXTO_HOST = getenv("MIXTO_HOST")
MIXTO_API_KEY = getenv("MIXTO_API_KEY")


class MissingRequired(Exception):
    """Missing params"""

    pass


class BadResponse(Exception):
    """Bad response from Mixto API"""

    pass


class MixtoLite:
    def __init__(
        self, host: Union[None, str] = None, api_key: Union[None, str] = None
    ) -> None:
        super().__init__()
        self.host = host
        self.api_key = api_key
        self.workspace_id = None
        self.status = 0
        self.commit_type = "tool"

        # if envars are set, always use those values
        if MIXTO_HOST is not None and self.host is None:
            self.host = MIXTO_HOST
        if MIXTO_API_KEY is not None and self.api_key is None:
            self.api_key = MIXTO_API_KEY

        # if host or apikey is not available, read config file
        if self.host == None or self.api_key == None:
            try:
                conf_path = str(Path().home() / ".mixto.json")
                with open(conf_path) as f:
                    j = json.loads(f.read())
                    self.host = j["host"]
                    self.api_key = j["api_key"]
                    self.workspace_id = j["workspace_id"]
            except:
                print("Cannot read mixto config file")
                raise

    def MakeRequest(
        self,
        method: str,
        uri: str,
        body: dict = {},
        query: dict = {},
        isJSON: bool = True,
    ):
        """Generic method helpful in extending this lib for other Mixto
        API calls. Refer to Mixto docs for all available API endpoints.

        Args:
            method (str): Request method
            uri (str): Mixto URI.
            body (dict, optional): Body. Defaults to {}.
            query (dict, optional): Query params. Defaults to {}.
            isJSON (bool, optional): If the response is of type JSON. Defaults to True.

        Raises:
            BadResponse: [description]
            BadResponse: [description]

        Returns:
            [type]: [description]
        """
        url = urljoin(str(self.host), uri)
        q = ""
        if query:
            q = "?" + urlencode(query)

        if self.api_key is None:
            raise AttributeError("api_key not found")

        req = Request(
            method=method.upper(),
            url=url + q,
            data=json.dumps(body).encode(),
            headers={
                "x-api-key": self.api_key,
                "user-agent": "mixto-lite-py",
            },
        )
        if body:
            req.add_header("Content-Type", "application/json")
        try:
            res = urlopen(req)
            body = res.read().decode()
            self.status = res.getcode()
            if self.status > 300:
                raise BadResponse(self.status, res)
            if isJSON:
                return json.loads(str(body))
            else:
                return body
        except HTTPError as e:
            raise BadResponse(e.code, e.read())

    def AddCommit(self, data: str, entry_id: str, title: str = "", optional: dict = {}):
        """Add/commit data to an entry. This is the primary functionality of
        an integration

        Args:
            data (str): Data to add
            entry_id (str, optional): Entry ID. Will use MIXTO_ENTRY_ID as primary. Defaults to None.
            title (str, optional): Title for commit. Defaults to "Untitled".
            optional (dict, optional): Optional dict to add to request body.

        Raises:
            MissingRequired: If entry id is missing

        Returns:
            dict: Commit added response
        """
        if MIXTO_ENTRY_ID is None and entry_id is None:
            raise MissingRequired("Entry id is missing")

        e_id = MIXTO_ENTRY_ID if MIXTO_ENTRY_ID else entry_id
        body = {
            "data": data,
            "workspace_id": self.workspace_id,
            "entry_id": e_id,
            "commit_type": self.commit_type,
            "title": title,
        }

        if len(optional) > 0:
            body = body | optional

        r = self.MakeRequest(
            "POST",
            "/api/v1/commit",
            body,
        )
        return r

    def GetWorkspaces(self) -> List[Dict[str, str]]:
        """Get all workspaces information and stats

        Returns:
            List[Dict[str, str]]: Array of workspace items
        """
        return self.MakeRequest("GET", "/api/v1/workspace")["data"]

    def GetEntryIDs(self, include_commits: bool = False) -> List[Dict[str, str]]:
        """Get all entry ids filtered by the current workspace

        Returns:
            List[Dict[str, str]]: List of entry ids
        """
        # get all entries
        entries = self.MakeRequest(
            "POST",
            "/api/v1/workspace",
            {"workspace_id": self.workspace_id, "include_commits": include_commits},
        )["data"]["entries"]
        # filter workspaces by current workspace
        return entries

    def GetCommitData(self, commit_id: str) -> str:
        """Get data for a commit by commit_id

        Args:
            commit_id (str): A valid commit_id

        Raises:
            ValueError: If no commit data is found

        Returns:
            dict: Commit added response
        """
        query = """query q($commit_id: uuid = "") {
            commit: mixto_commits_by_pk(commit_id: $commit_id) {
                data
            }
        }"""

        commit_data = self.MakeRequest(
            "POST",
            "/api/v1/gql",
            {"query": query, "variables": {"commit_id": commit_id}},
        )

        if "data" not in commit_data:
            raise ValueError("commit data not found")
        elif "commit" not in commit_data["data"]:
            raise ValueError("commit data not found")
        elif "data" not in commit_data["data"]["commit"]:
            raise ValueError("commit data not found")

        return commit_data["data"]["commit_id"]["data"]

    def GraphQL(
        self, query: str, variables: Union[Dict[str, Any], None] = None
    ) -> Dict[str, Any]:
        """Make a graphql request

        Args:
            query (str): GQL query string
            variables (Union[Dict[str, Any], None], optional): GQL variables. Defaults to None.

        Raises:
            ValueError: If the data key is not found in the response

        Returns:
            Dict[str, Any]: GQL response
        """
        body: Dict[str, Any] = {"query": query}
        if variables is not None:
            body["variables"] = variables
        resp = self.MakeRequest("POST", "/api/v1/gql", body=body)

        if "data" not in resp:
            raise ValueError(resp)

        return resp["data"]
